### DISCUSSION TITLE
v2.2.1-B - Community Builds Thread

### DISCUSSION THREAD

**Discussion by gianni-rosato** [gianni-rosato/svt-av1-psy#79] [Show and tell]
# Community Builds Thread

This is a place for the community to share unofficial tools not affiliated with the project — mainly consisting of binaries compiled by community members.

## Trust

:warning: **Community-built binaries are not affiliated with the SVT-AV1-PSY project & are to be used at your own risk.** If you encounter bugs while using a community-built binary, ensure that you double-check with the person responsible for providing you with the binary before submitting an issue report.

## Architecture

When downloading pre-compiled binaries, you might see AVX, AVX2, AVX-512, x86-64-v3, etc. If you don't know exactly what ISA extensions your CPU supports, here is a chart to help you quickly understand your hardware's support:

### Not Vendor-Specific
```c
x86-64    : Processors lacking SSE4.2 and newer features
x86-64-v2 : Processors with features up to and including SSE4.2
x86-64-v3 : Processors with features up to and including AVX and/or AVX2
x86-64-v4 : Processors with features up to and including AVX512
```

ISA extensions like AVX, AVX2, & AVX512 are for *vector processing* which SVT-AV1-PSY relies heavily on for its fast multithreaded performance. Below are some helpful charts to help you narrow down the best options if you have popular x64 hardware:

### AMD
```ocaml
Generation                                             Vectors  Build
Zen   (1xxx)                                           AVX2     x86-64-v3 / znver1
Zen+  (1xxxAF-2xxx-3xxxG/H/U/C)                        AVX2     x86-64-v3 / znver1
Zen2  (3xxx-4xxxS/G/H/U-5700U-5500U-5300U-7520U-7320U) AVX2     x86-64-v3 / znver2
Zen3  (5xxx-7330U-7530U-7730U)                         AVX2     x86-64-v3 / znver3
Zen3+ (6xxx-7xxxH/U)                                   AVX2     x86-64-v3 / znver3
Zen4  (7xxx-8xxxG)                                     AVX512   x86-64-v4 / znver4
```
### Intel (Desktop)
```js
Generation                Vectors  Build
Haswell     (4th gen)     AVX2     x86-64-v3 / haswell
Broadwell   (5th gen)     AVX2     x86-64-v3 / broadwell
Skylake     (6th gen)     AVX2     x86-64-v3 / skylake
Kaby Lake   (7th gen)     AVX2     x86-64-v3 / skylake
Coffee Lake (8-9th gen)   AVX2     x86-64-v3 / skylake
Comet Lake  (10th gen)    AVX2     x86-64-v3 / skylake
Rocket Lake (11th gen)    AVX512   x86-64-v4 / rocketlake
Alder Lake  (12th gen)    AVX2     x86-64-v3 / alderlake
Raptor Lake (13-14th gen) AVX2     x86-64-v3 / raptorlake
```

Included below is even more information about what is available when specifying `-march` & `-mtune` on x64 CPUs:

Known valid x64 arguments for `-march=`:
> i386 i486 i586 pentium lakemont pentium-mmx winchip-c6 winchip2 c3 samuel-2 c3-2 nehemiah c7 esther i686 pentiumpro pentium2 pentium3 pentium3m pentium-m pentium4 pentium4m prescott nocona core2 nehalem corei7 westmere sandybridge corei7-avx ivybridge core-avx-i haswell core-avx2 broadwell skylake skylake-avx512 cannonlake icelake-client rocketlake icelake-server cascadelake tigerlake cooperlake sapphirerapids emeraldrapids alderlake raptorlake meteorlake graniterapids graniterapids-d arrowlake arrowlake-s lunarlake pantherlake bonnell atom silvermont slm goldmont goldmont-plus tremont gracemont sierraforest grandridge clearwaterforest knl knm intel geode k6 k6-2 k6-3 athlon athlon-tbird athlon-4 athlon-xp athlon-mp x86-64 x86-64-v2 x86-64-v3 x86-64-v4 eden-x2 nano nano-1000 nano-2000 nano-3000 nano-x2 eden-x4 nano-x4 lujiazui yongfeng k8 k8-sse3 opteron opteron-sse3 athlon64 athlon64-sse3 athlon-fx amdfam10 barcelona bdver1 bdver2 bdver3 bdver4 znver1 znver2 znver3 znver4 znver5 btver1 btver2 generic native

Known valid x64 arguments for `-mtune=`:
> generic i386 i486 pentium lakemont pentiumpro pentium4 nocona core2 nehalem sandybridge haswell bonnell silvermont goldmont goldmont-plus tremont sierraforest grandridge clearwaterforest knl knm skylake skylake-avx512 cannonlake icelake-client icelake-server cascadelake tigerlake cooperlake sapphirerapids alderlake rocketlake graniterapids graniterapids-d arrowlake arrowlake-s pantherlake intel lujiazui yongfeng geode k6 athlon k8 amdfam10 bdver1 bdver2 bdver3 bdver4 btver1 btver2 znver1 znver2 znver3 znver4 znver5

- Note: `-march=foo` implies `-mtune=foo` unless you also specify a different `-mtune`. This is one reason why using `-march` is better than just enabling options like `-mavx` without doing anything about tuning.

## Antivirus

🛑 Be wary of antivirus software on Windows detecting EXEs distributed here as malicious software. While they may not always legitimately be malicious, it is important to maintain a healthy level of skepticism when running code that someone else has compiled.

**Comment by Uranite** [comment]
SvtAv1EncApp v2.2.1-B for Windows
```
cmake -B svt_build -T ClangCL -DBUILD_SHARED_LIBS=OFF -DENABLE_AVX512=$AVX512Flag -DCMAKE_CXX_FLAGS_RELEASE="$BaseFlags -march=$Arch" -DCMAKE_C_FLAGS_RELEASE="$BaseFlags -march=$Arch"
```
$BaseFlags = "/DNDEBUG /clang:-O2 -flto" (same as Itachi Uchiha's build)
Includes `x86-64, x86-64-v2, x86-64-v3, x86-64-v4, znver1, znver2, znver4, haswell, broadwell, skylake, rocketlake, alderlake, raptorlake`
AVX512 is on for `x86-64-v4, znver4, rocketlake`

[SvtAv1EncApp-windows-all-arch-but-znver3.zip](https://github.com/user-attachments/files/17325522/SvtAv1EncApp-windows-all-arch-but-znver3.zip)

**Reply by tuanden0** [reply]
cry in AMD 5950X

**Reply by Uranite** [reply]
use znver2 because znver3 is actually slower, that's why znver3 isn't included

**Reply by tuanden0** [reply]
is znver2 faster than x86-64-v3 for AMD 5000?

**Reply by Uranite** [reply]
> is znver2 faster than x86-64-v3 for AMD 5000?

yeah it's faster

**Reply by tuanden0** [reply]
Thank you very much

**Reply by ItachiUchiha-IU** [reply]
For more infos, findings and discussions about the builds (the znver3 case, the parameters used and tested, w10-11, clang 17-19, ninja, **_PGO_**, **_BOLT_**), check the [Community Builds Thread for v2.2.1-A](https://github.com/gianni-rosato/svt-av1-psy/discussions/69)

Everyone is especially invited to contribuite about PGO and BOLT on Windows!!

**Reply by gitoss** [reply]
> Everyone is especially invited to contribuite about PGO and BOLT on Windows!!

I'm not seeing anything about BOLT on Windows - it's not provided with the LLVM 19 package for Windws, the git readme states "BOLT operates on X86-64 and AArch64 ELF binaries" and the BOLT description on their website sounds more like POSIX, too.

As far as I understand it, middle-end context sensitive (IR CS) PGO should/could do a good part of what the context-sensivite BOLT is designed to do. https://github.com/llvm/llvm-project/issues/56274#issuecomment-1407117774

**Reply by gitoss** [reply]
For testing purposes, here's an optimized znver4 binary using fast-math plus 2-stage cs-ir pgo.

[SvtAv1EncApp_llvm19-csirpgo-fastmath_znver4_v2.2.1-B.zip](https://github.com/user-attachments/files/17334762/SvtAv1EncApp_llvm19-csirpgo-fastmath_znver4_v2.2.1-B.zip)

I've profiled using presets 5 to 3 with a 10-bit fhd non-anime video - i.e. other videos might not match the profile in the same way. Please let me/us know if there are any speed regressions (because of pgo) or artifacts (because of fast-math).

For building for other targets see the current pgo thread https://github.com/gianni-rosato/svt-av1-psy/discussions/69 - the whole pgo build process could probably be automated using cmake, either using vs' default llvm 17 or even the llvm 19 add-in.

**Reply by n00mkrad** [reply]
@Uranite Any chance you could compile with DoVi enabled?

**Reply by gitoss** [reply]
> @Uranite Any chance you could compile with DoVi enabled?

For libdovi to work without problems, it would be best if svt-av1-psy would pull the (few) changes from this fork: https://github.com/Patman86/SVT-AV1-Mod-by-Patman/tree/PSY-PMod

**Reply by gianni-rosato** [reply]
Urge Patman to make a PR and we will review the relevant changes.

**Reply by Uranite** [reply]
> @Uranite Any chance you could compile with DoVi enabled?

I could but the dovi dll wont be static

**Reply by Uranite** [reply]
> > @Uranite Any chance you could compile with DoVi enabled?
> 
> For libdovi to work without problems, it would be best if svt-av1-psy would pull the (few) changes from this fork: https://github.com/Patman86/SVT-AV1-Mod-by-Patman/tree/PSY-PMod

Patman's changes are no longer necessary with #68
Edit: It _should_ no longer be necessary, it builds fine but I didn't test encoding or playback

**Reply by shuniichan** [reply]
So znver3 still broken in this build like in A release?

**Comment by Uranite** [comment]
SvtAv1EncApp v2.2.1-B for Termux
Includes `arm` and `aarch64`
pick one according to your arch and install the .deb file via `dpkg -i`
afterwards you can encode via ffmpeg
[svt-av1-psy-termux.zip](https://github.com/user-attachments/files/17325648/svt-av1-psy-termux.zip)
